### PR TITLE
Block new incoming requests once close has been called

### PR DESCRIPTION
**Pull request by delvedor**
Currently, if a user calls `fastify.close` and the server is under a heavy load, the server will not close until there is an open TCP connection, so every new request will execute the entire request lifecycle.

With this change once `fastify.close` is invoked, Fastify will automatically start to respond with a `503` error, closing every new connection as soon as possible.

There are not sensible changes in the benchmarks.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by delvedor** [comment]
> I'm -1 with this feature (from my point of view this is a breaking behaviour change).

Actually, it can be interpreted as a bugfix.
Close is a LIFO queue of `onClose` handlers, once a user calls `close` it will activate that queue, closing database connections for example. This will lead to many errors inside the various route handlers, which is not very nice.
With this change, we are actually protecting the server from all that kind of errors.

**Comment by nwoltman** [comment]
> Close is a LIFO queue of `onClose` handlers

Really? I assumed it was FIFO. So in this code:

```js
fastify.onClose(async () => {
  // This is called second
})

fastify.onClose(async () => {
  // And this is called first
})
```
?

**Comment by allevo** [comment]
I don't understand. I expect that the HTTP server is closed before **any** other dependencies. Is this the currently behaviour ?

**Comment by delvedor** [comment]
> Really? I assumed it was FIFO.

@nwoltman yes. The reason is because the last plugin you have have registered should be the first you must close.

**Comment by delvedor** [comment]
@nwoltman done. But the problem persist, because of the keep alive connections as @mcollina pointed out.

**Comment by nwoltman** [comment]
@delvedor You're right, I forgot about the keep-alive connections. Would it help to include a `Connection: close` header in the 503 response?

**Comment by delvedor** [comment]
@nwoltman it should be the [default](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection), but it will not hurt ;)

**Comment by nwoltman** [comment]
`close` is the default for HTTP/1.0 requests, but modern browsers use HTTP/1.1 where the default is to use a `keep-alive` connection.

The changes LGTM, although I think @mcollina has [a good point](https://github.com/fastify/fastify/pull/886#pullrequestreview-112252474) about testing with something that uses keep-alive connections instead of `.inject()`.

**Comment by delvedor** [comment]
I'm trying to write a test with keep alive, but I'm having some trouble to get it working.
Any suggestion?
https://gist.github.com/delvedor/558a1569d7db6e2c0c9e3870eb96fd33

**Comment by mcollina** [comment]
@delvedor you should verify it's reusing the same socket. I don't think it is, as you are not resuming the response. Also, add a `maxSocket: 1` option to the agent.

File: fastify.js [hunk 1]
```diff
@@ -200,6 +204,16 @@ function build (options) {
   return fastify
 
   function routeHandler (req, res, params, context) {
+    if (closing === true) {
+      res.writeHead(503, {
+        'Content-Type': 'application/json',
+        'Content-Length': '80',
```
Comment (StarpTech) [inline fastify.js]: Could you calculate it based on the payload [as with 404](https://github.com/fastify/fastify/blob/master/fastify.js#L708) ?

**Comment by StarpTech** [comment]
@delvedor I'm trying to reproduce it on my machine.

File: fastify.js (hunk 1 above)
Comment (delvedor) [reply fastify.js]: That `80` is the length of the json response ;)

Comment (StarpTech) [reply fastify.js]: I know. It's better to write `payload.length`

Comment (delvedor) [reply fastify.js]: **It is** the payload length.
We can do the same at the line you have pointed pointed out, since the payload will be always the same.

**Comment by StarpTech** [comment]
@delvedor can't reproduce it on local Node 6.13.1 I think the timeouts aren't solid try to increase them.

**Comment by delvedor** [comment]
@StarpTech one of the two AppVeyor logs shows
`error is not logged because promise was fulfilled with undefined but response was sent before promise resolution not ok Request timed out`.
Which is not a test we are running in this branch. Do you know why?

**Comment by StarpTech** [comment]
@delvedor no idea. Try to rebuild.

**Comment by mcollina** [comment]
done.

**Comment by mcollina** [comment]
You'll need to increaase that timeout: https://ci.appveyor.com/project/mcollina/fastify/build/job/x2ml3eliuojq3au2#L41.

**Comment by StarpTech** [comment]
@delvedor it's correct

image

**Comment by StarpTech** [comment]
@delvedor the promise test was there because Appveyor test the PR + the merge result.

**Comment by delvedor** [comment]
I don't know why the test is randomly failing, I'll investigate more later.

**Comment by delvedor** [comment]
@StarpTech now the test *should* be fixed.
But AppVeyor is still failing on the PR + the merge, can you give it a check?

**Comment by StarpTech** [comment]
I'm investigating

**Comment by StarpTech** [comment]
@delvedor I think this can be fixed with a higher timeout of `sget` calls. Should I create a PR ? I can't reproduce it locally.

**Comment by StarpTech** [comment]
or you merge master and adjust the number to e.g 500.

**Comment by delvedor** [comment]
@StarpTech open a pr :)

**Comment by StarpTech** [comment]
@delvedor https://github.com/fastify/fastify/pull/888

**Comment by delvedor** [comment]
I've tried with no timeout, 150ms, 250ms, 500ms and 1000ms.
I don't think it is a problem of timing.
I can't reproduce the problem on my machine. Any ideas?

**Comment by StarpTech** [comment]
I could reproduce it locally and I could retrace it until:

* [callback](https://github.com/fastify/fastify/blob/block-requests-on-close/fastify.js#L117) or `close` event of server instance is never called

when I set `fastify.server.setTimeout(200)` the callback is executed but the same error appears.

```
test/close.test.js .................................... 2/3 30s
  Should return 503 while closing - keep alive
  not ok (unnamed test)
    stack: |-
      Error: read ECONNRESET
          at exports._errnoException (util.js:1020:11)
          at TCP.onread (net.js:580:26)
    message: read ECONNRESET
    code: ECONNRESET
    errno: ECONNRESET
    syscall: read

total ................................................. 2/3
```

**Comment by StarpTech** [comment]
@delvedor just as a hint. Hapi graceful shutdown keepAlive connections very differently. They destroy them manually.

* https://github.com/hapijs/hapi/blob/f9df1dd8d37a6fe5be8dd7f1fc3c7b8213341199/lib/core.js#L391
* https://github.com/hapijs/hapi/blob/916808fc18c26353126e814d9d4ef84f3896c7a2/test/core.js#L842

**Comment by mcollina** [comment]
They do not call `.destroy()` (which is not graceful), but rather they end them manually (https://github.com/hapijs/hapi/blob/f9df1dd8d37a6fe5be8dd7f1fc3c7b8213341199/lib/core.js#L409-L412).

I'm 👍 with that approach and it might be better than this one.

**Comment by delvedor** [comment]
Hapi is using a Set, which means that you should add the socket to the set for every connection and remove the socket from the Set once the connection has been closed.
I fear it is too slow.

**Comment by StarpTech** [comment]
@delvedor performance of what? The request / response shouldn't be effected.

**Comment by delvedor** [comment]
@StarpTech for every new connection you should update a `Set` and every time that a connection will be closed you should again update that `Set`. This is overhead that I don't want to pay unless I don't have to.

**Comment by StarpTech** [comment]
@delvedor you are worried about to maintain a set? What's the alternative to close all pending connections? From my understanding, the problem is that `server.close` does not clean up the connections?

**Comment by delvedor** [comment]
I'm worried about the **cost** of maintaining a `Set`.

Regarding `server.close`, this is the behavior ([documentation](https://nodejs.org/dist/latest-v8.x/docs/api/net.html#net_server_close_callback)):
> Stops the server from accepting new connections and keeps existing connections. This function is asynchronous, the server is finally closed when all connections are ended

**Comment by StarpTech** [comment]
@delvedor https://github.com/hunterloftis/stoppable from thread https://github.com/hunterloftis/stoppable

**Comment by StarpTech** [comment]
There is also an interesting thread about it https://github.com/nodejs/node/issues/2642 Node.js does not handle existing, idle connections. We should definitely support it.

**Comment by delvedor** [comment]
@StarpTech I know, I've already read that issue before open this pr.
The problem is the same, you are keeping a `Map`, calling `.set` and `.delete` very often.
That is pure overhead. I'm looking for a solution with the least overhead.

What I don't understand it's why the test are failing in travis but not in my local machine.

**Comment by StarpTech** [comment]
They are failing really randomly. Sometimes it works sometimes not. Try to run the single test in isolation.

**Comment by delvedor** [comment]
@mcollina can you check if the usage of `autocannon` is correct?

**Comment by delvedor** [comment]
Done, I've also added another check.
What is the difference?

File: test/close-pipelining.test.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      connections: 1,
+      amount: 3
+    })
+
+    const codes = [200, 503, 'error']
+    instance.on('response', (client, statusCode) => {
+      t.strictEqual(codes.shift(), statusCode)
+    })
+
+    instance.on('done', () => t.pass())
```
Comment (allevo) [inline test/close-pipelining.test.js]: Can you use t.done? It is more clear

File: fastify.js [hunk 3]
```diff
@@ -109,12 +110,15 @@ function build (options) {
     server = http.createServer(httpHandler)
   }
 
-  fastify.onClose((instance, done) => {
-    if (listening) {
-      instance.server.close(done)
-    } else {
-      done(null)
-    }
+  app.once('preReady', () => {
```
Comment (allevo) [inline fastify.js]: Maybe overwriting avvio ready method fits better here. Or not?

Comment (delvedor) [reply fastify.js]: Can you rephrase?

Comment (allevo) [reply fastify.js]: Putting the `onClose` handler inside `preReady` event doesn't ensure that the callback is invoked before any others. It does only if the `preReady` callbacks are fired in FIFO.

We use

https://github.com/mcollina/avvio/blob/a957201b0b33facf06687a5ba8e52c82a46b1302/boot.js#L13

and creating `ready` method inside fastify.

The new method forces the add `onClose` handler as the last, so the first one to be  invoked.

**Comment by mcollina** [comment]
Green CI! 🎉
